greeter/README.md

```diff
@@ -1,21 +1,510 @@
+```json
+{
+  "timestamp": "2019-10-01T18:17:07.730Z",
+  "language": "zh",
+  "greeting": "Diana，你好！"
+}
+```
+
+
+#### Test 3: Set the Language to French `fr`
+Let's try it one more time with language `fr`.  Since there is no greeter for French, so it will fall back to use English.  The response will be as follows:
+```json
+{
+  "timestamp": "2019-10-01T18:26:45.742Z",
+  "language": "fr",
+  "greeting": "Hello, Diana!"
+}
+```
+
+
+## Part 2: Add a French Greeter  
+We are now going to create a new French greeter class which extends a certain interface defined by the extension point. Then it has to register itself against that extension point.
+
+
+### Step 1: Creater French Greeter
+1. In `src` folder, create a file called `greeter-fr.ts`. 
+2. Add the required imports.
+    ```ts
+    import {bind, config} from '@loopback/context';
+    import {asGreeter, Greeter} from '@loopback/example-greeter-extension';
+    ```
+
+3. Add the `FrenchGreeter` class. 
+    ```ts
+    /**
+     * A greeter implementation for French.
+     */
+    @bind(asGreeter)
+    export class FrenchGreeter implements Greeter {
+    language = 'fr';
+
+    constructor(
+        /**
+         * Inject the configuration for FrenchGreeter
+         */
+        @config()
+        private options: FrenchGreeterOptions = {nameFirst: true},
+    ) {}
+
+    greet(name: string) {
+        if (this.options && this.options.nameFirst === false) {
+        return `Bonjour，${name}！`;
+        }
+        return `${name}, Bonjour`;
+    }
+    }
+    ```
+2. Add the `FrenchGreeterOptions` interface.
+
+    ```ts
+    /**
+     * Options for the Frech greeter
+     */
+    export interface FrenchGreeterOptions {
+        // Name first, default to `true`
+        nameFirst: boolean;
+    }
+    ```
+
+
+
+### Step 2: Bind the FrenchGreeter in the Application
+
+In `src/application.ts`, inside the constructor, add the line below: 
+```ts
+ this.component(GreetingComponent);
+ // add this line 
+ this.add(createBindingFromClass(FrenchGreeter));
+```
+
+Add the following import statements:
+```ts
+import {ApplicationConfig, createBindingFromClass} from '@loopback/core'; //Change this line
+import {FrenchGreeter} from './greeter-fr'; //Add this line
+```
+
+### Step 3: Try Out the New Greeter
+Let's try out the new French Greeter.  
+1. Stop the application by entering Ctrl+C.
+2. Restart the application again by running `npm start`. 
+3. Go to API Explorer, http://localhost:3000/explorer.
+4. Under the `GET /greet/{name}` endpoint, enter `fr` as the language as shown below. 
+
+    ![Test French Greeter](images/french-greeter-test.png)
+
+5. In the response body section, you should be able to see the greeting in French.
+    ```json
+    {
+    "timestamp": "2019-10-01T18:47:37.096Z",
+    "language": "fr",
+    "greeting": "Diana, Bonjour"
+    }
+    ```
+
+
+## Part 3: Enable Caching
+
+_TODO: what we want to do in this part_ 
```

Comment:
Extending the functionality of endpoint(s) by applying **interceptor** and **observer**.

#### Concepts

- Interceptors: are reusable functions to provide aspect-oriented logic around method invocations. 

- Observer: A LoopBack application has its own life cycles at runtime. The transition of states is controlled by two functions: `start()` and `stop()`. Observer allows artifacts to participate in the life cycles.

*Concepts are very abstract, next let's explain what is interceptor and observer by adding a feature called cache service to the greeter endpoint. Let's start with listing our requirements.*

#### Requirements

- Requirement 1: The greeting msg should be cached on the server side so that next time when a request with same greeting name and language sent to the app, it returns the cached result instead of calculating the result again.
- Requirement 2: The caching service should have an expiration time for the cached results. E.g.  a result should be regenerated after 10 mins.
- Requirement 3: The caching service should sweep the cached result after a certain interval to clean up the expired result.

#### Implementation

- Interceptor could do requirement 1 in a way that every time before the greeter function gets invoked, the cache interceptor retrieves the cached result first. If found, then return without invoking the method.
- Observer could do 2&3. Create a cache service which sets up a timer when the app runs, sweeps the cache after a set period of time. The observer controls the start and stop of the service.

#### Summary

The interceptor, service and observer could be contributed by from different modules, or packaged together as a component, all depends on your choice. You just bind them to the app without modifying the business logic of your endpoint.

Back to the concepts, now audience should have a more through understanding of how to "provide aspect-oriented logic around method invocations" and how to "allow artifacts to participate in the life cycles".